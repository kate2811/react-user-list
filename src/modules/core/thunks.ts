import actions from './actions'
import { User, UserData } from './types'
import { State } from '../../store'
import { v4 } from 'uuid'

function updateUserList(getState: () => State) {
  const currentUsers = getState().core.userList
  localStorage.setItem('users', JSON.stringify(currentUsers))
}

export function removeUser(id: string) {
  return function(dispatch: any, getState: () => State) {
    if (window.confirm('Do you really want to remove this user?')) {
      dispatch(actions.removeUser(id))
      updateUserList(getState)
    }
  }
}

export function addUser(userData: UserData) {
  return function(dispatch: any, getState: () => State) {
    const id = v4()
    const currentUsers = getState().core.userList
    const allUsers = [...currentUsers, { ...userData, id }]
    localStorage.setItem('users', JSON.stringify(allUsers))
    dispatch(actions.addUser({ ...userData, id }))
  }
}

export function editUser(userData: User) {
  return function(dispatch: any, getState: () => State) {
    dispatch(actions.editUser(userData))
    updateUserList(getState)
  }
}

export function loadUserList() {
  return function(dispatch: any) {
    dispatch(actions.loadUserList())
    new Promise((resolve) => {
      setTimeout(() => resolve(JSON.parse(localStorage.getItem('users') as string) || []), 2000)
    }).then((result) => dispatch(actions.loadUserListSuccess(result)))
  }
}
