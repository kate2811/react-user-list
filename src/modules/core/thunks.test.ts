import configureStore from 'redux-mock-store'
import { addUser, removeUser, editUser, loadUserList } from './thunks'
import thunk from 'redux-thunk'
import { v4 } from 'uuid'

jest.mock('uuid', () => ({
  v4: () => '1'
}))

const history = {
  push: jest.fn()
}
const storage = {
  setItem: jest.fn(),
  getItem: jest.fn(() => JSON.stringify([{ name: 'Elena', surname: 'Ivanova', email: 'jj', age: 1 }]))
}

const middlewares: any[] = [thunk.withExtraArgument({ history, storage })]
const mockStore = configureStore(middlewares)
const mockUserData = { name: 'Elena', surname: 'Ivanova', email: 'jj', age: 1 }

describe('thunks', () => {
  const initialState = { core: { userList: [{ name: 'Kate', surname: 'Pavlova', email: 'jj', age: 1, id: '3' }] } }
  const store = mockStore(initialState)
  const currentUsers = initialState.core.userList

  beforeEach(() => {
    jest.clearAllMocks()
    store.clearActions()
  })

  it('addUser', () => {
    store.dispatch(addUser(mockUserData))
    const id = v4()
    const allUsers = [...initialState.core.userList, { ...mockUserData, id }]

    expect(store.getActions()).toMatchSnapshot()
    expect(history.push).toHaveBeenCalledTimes(1)
    expect(history.push).toHaveBeenCalledWith('/')
    expect(storage.setItem).toHaveBeenCalledTimes(1)
    expect(storage.setItem).toHaveBeenCalledWith('users', JSON.stringify(allUsers))
  })

  it('removeUser confirmed', () => {
    window.confirm = jest.fn(() => true)
    store.dispatch(removeUser('1'))

    expect(store.getActions()).toMatchSnapshot()
    expect(storage.setItem).toHaveBeenCalledTimes(1)
    expect(storage.setItem).toHaveBeenCalledWith('users', JSON.stringify(currentUsers))
  })

  it('removeUser uncomfirmed', () => {
    window.confirm = jest.fn(() => false)
    store.dispatch(removeUser('1'))

    expect(store.getActions()).toMatchSnapshot()
    expect(store.getActions()).toHaveLength(0)
    expect(storage.setItem).toHaveBeenCalledTimes(0)
  })

  it('editUser', () => {
    store.dispatch(editUser({ ...mockUserData, id: '2' }))

    expect(store.getActions()).toMatchSnapshot()
    expect(storage.setItem).toHaveBeenCalledTimes(1)
    expect(storage.setItem).toHaveBeenCalledWith('users', JSON.stringify(currentUsers))
  })

  it('loadUserList', async () => {
    await store.dispatch(loadUserList())
    expect(store.getActions()).toMatchSnapshot()
    expect(storage.getItem).toHaveBeenCalledTimes(1)
  })
})
