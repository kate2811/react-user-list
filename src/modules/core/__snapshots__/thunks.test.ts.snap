// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`thunks addUser 1`] = `
Array [
  Object {
    "payload": Object {
      "age": 1,
      "email": "jj",
      "id": "1",
      "name": "Elena",
      "surname": "Ivanova",
    },
    "type": "[Core] Add a new user",
  },
]
`;

exports[`thunks editUser 1`] = `
Array [
  Object {
    "payload": Object {
      "age": 1,
      "email": "jj",
      "id": "2",
      "name": "Elena",
      "surname": "Ivanova",
    },
    "type": "[Core] Edit user",
  },
]
`;

exports[`thunks loadUserList 1`] = `
Array [
  Object {
    "payload": undefined,
    "type": "[Core] Load users list",
  },
  Object {
    "payload": Array [
      Object {
        "age": 1,
        "email": "jj",
        "name": "Elena",
        "surname": "Ivanova",
      },
    ],
    "type": "[Core] Load users list success",
  },
]
`;

exports[`thunks removeUser confirmed 1`] = `
Array [
  Object {
    "payload": "1",
    "type": "[Core] Remove user",
  },
]
`;

exports[`thunks removeUser editUser 1`] = `
Array [
  Object {
    "payload": Object {
      "age": 1,
      "email": "jj",
      "id": "2",
      "name": "Elena",
      "surname": "Ivanova",
    },
    "type": "[Core] Edit user",
  },
]
`;

exports[`thunks removeUser uncomfirmed 1`] = `Array []`;
